main.cpp

```diff
@@ -38,8 +39,271 @@
+    size_t insert_begin = std::rand() % in1_size;
+    size_t insert_size = (std::rand() % max_insert_size) + 1;
+    size_t in2_begin = std::rand() % (in2_size - insert_size + 1);
+
+    std::memcpy(out, in1, insert_begin);
+    std::memcpy(out + insert_begin, in2 + in2_begin, insert_size);
+    std::memcpy(out + insert_begin + insert_size, in1 + insert_begin,
+               in1_size - insert_begin);
+
+    return in1_size + insert_size;
+}
+
+static size_t overwrite_part(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                           size_t in2_size, uint8_t *out, size_t max_out_size)
+{
+    if (in1_size > max_out_size || in1_size == 0)
+        return 0;
+
+    size_t overwrite_begin = std::rand() % in1_size;
+    size_t overwrite_size = (std::rand() % (in1_size - overwrite_begin)) + 1;
+    overwrite_size = std::min(overwrite_size, in2_size);
+    size_t in2_begin = std::rand() % (in2_size - overwrite_size + 1);
+
+    std::memcpy(out, in1, in1_size);
+    std::memcpy(out + overwrite_begin, in2 + in2_begin, overwrite_size);
+
+    return in1_size;
+}
+
+size_t cross_over(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                size_t in2_size, uint8_t *out, size_t max_out_size, unsigned seed)
+{
+    std::srand(seed);
+    if (std::rand() % 2)
+        return insert_part(in1, in1_size, in2, in2_size, out, max_out_size);
+    return overwrite_part(in1, in1_size, in2, in2_size, out, max_out_size);
+}
+
+size_t LLVMFuzzerCustomCrossOver(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                               size_t in2_size, uint8_t *out, size_t max_out_size,
+                               unsigned seed)
+{
+    if (in1_size < 9 || in2_size < 9)
+        return 0;
+
+    // Interpret fuzz inputs as strings
+    std::string input1(reinterpret_cast<const char*>(in1), in1_size);
+    std::string input2(reinterpret_cast<const char*>(in2), in2_size);
+
+    // Attempt to bech32 decode the inputs
+    bech32::DecodeResult result1 = bech32::Decode(input1, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (result1.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed
+        return 0;
+    }
+
+    bech32::DecodeResult result2 = bech32::Decode(input2, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (result2.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed
+        return 0;
+    }
+
+    std::string hrp1 = result1.hrp;
+    std::string hrp2 = result2.hrp;
+    std::vector<uint8_t> data1 = result1.data;
+    std::vector<uint8_t> data2 = result2.data;
+
+    std::srand(seed);
+    std::string out_hrp;
+    std::vector<uint8_t> out_data;
+
+    if (std::rand() % 2) {
+        // Cross-over the HRP
+        out_data = data1;
+
+        // Convert strings to vectors for cross_over function
+        std::vector<uint8_t> hrp1_vec(hrp1.begin(), hrp1.end());
+        std::vector<uint8_t> hrp2_vec(hrp2.begin(), hrp2.end());
```

Comment:
Do the HRPs actually need to be converted?  Maybe we can just use `data()` and `size()` on the strings themselves.

We might also be able to do this for `out_hrp`.

Reply:
I removed the conversion and it's now using `std::span` for the cross_over.